### PR TITLE
MPI Dependency

### PR DESCRIPTION
**Pull request by sslattery**
Adds an optional MPI dependency to the build and open-mpi to the Travis CI. This PR has no code that uses MPI - just adds the library and test infrastructure.

### CODE REVIEW

**Comment by sslattery** [comment]
@junghans Gave this an initial try. Looks like its failing to link on the C++ standard library in Travis.

Can you see anything obvious I did wrong? Basically, I added openmpi to the environment and then set the compiler under the MPI wrapper to be whichever we are using (GCC or clang).

**Comment by junghans** [comment]
@sslattery setting `CXX` to `mpicxx` isn't the CMake way, let me see if I can fix it.

File: CMakeLists.txt [hunk 1]
```diff
@@ -64,20 +64,31 @@ if(Doxygen_FOUND)
...
 endif()
 
+option(Cabana_ENABLE_MPI "Build Cabana with MPI support" OFF)
+if( Cabana_ENABLE_MPI )
+  find_package(MPI)
+  if( NOT MPI_FOUND )
+    message(FATAL_ERROR "Cabana MPI support requested but MPI not found." )
+  endif()
+else()
+  set(MPI_CXX__LIBRARIES)
```
Comment (dalg24) [inline CMakeLists.txt]:
```suggestion
  set(MPI_CXX__LIBRARIES)
  set(MPI_CXX_LIBRARIES)
```

File: core/src/CMakeLists.txt [hunk 2]
```diff
@@ -17,6 +17,6 @@ install(FILES ${CMAKE_CURRENT_BINARY_DIR}/CabanaCore_config.hpp DESTINATION ${CM
 #-----------------------------------------------------------------------------
 
 add_library(cabanacore ${SOURCES} ${SOURCES} ${SOURCES_IMPL})
-target_include_directories(cabanacore PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
-target_link_libraries(cabanacore Kokkos::kokkos)
+target_include_directories(cabanacore PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR} ${MPI_CXX_INCLUDE_PATH})
+target_link_libraries(cabanacore Kokkos::kokkos ${MPI_CXX__LIBRARIES})
```
Comment (dalg24) [inline core/src/CMakeLists.txt]:
```suggestion
target_link_libraries(cabanacore Kokkos::kokkos ${MPI_CXX_LIBRARIES})
```

**Comment by junghans** [comment]
From which CMake version on is the `MPI::MPI_CXX` Target supported?

**Comment by junghans** [comment]
Good point, then we should use the imported target!

**Comment by sslattery** [comment]
So on Travis it looks like it is still using `g++` or `clang` to do the compiling. Does the way you have linked in the CMake supersede the use of `mpicxx`?

**Comment by sslattery** [comment]
OK just verified with myself that this works by injecting some MPI code into the unit tests to try it out. This is all I wanted to do on this PR. I'll work on package structure and then start on the communication PRs next. If both of you are fine with this please give a review and then we can merge. @junghans @dalg24 thanks for the help.

**Comment by sslattery** [comment]
Also once we merge I will update the [build instructions](https://github.com/ECP-copa/Cabana/wiki/Build-Instructions)

File: core/src/CMakeLists.txt [hunk 3]
```diff
@@ -19,4 +19,8 @@ install(FILES ${CMAKE_CURRENT_BINARY_DIR}/CabanaCore_config.hpp DESTINATION ${CM
 add_library(cabanacore ${SOURCES} ${SOURCES} ${SOURCES_IMPL})
 target_include_directories(cabanacore PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
 target_link_libraries(cabanacore Kokkos::kokkos)
+if(Cabana_ENABLE_MPI)
+  find_package(MPI REQUIRED)
```
Comment (dalg24) [inline core/src/CMakeLists.txt]: Not sure I'd put the `find_package(MPI)` here.  You might want to do that in the root directory `CMakeLists.txt` so you can register test with `${MPIEXEC} ${MPIEXEC_NUMPROC_FLAG} ...`

Also do not forget the `if (NOT MPI_CXX_FOUND) message(FATAL_ERROR ...`

Comment (junghans) [reply core/src/CMakeLists.txt]: The former is a good point as tests might not live in a subfolder!

The latter, I was thinking `REQUIRED` will check for all enabled languages, incl. `CXX`.

Comment (sslattery) [reply core/src/CMakeLists.txt]: @junghans Do you want to make these changes or should I?

Comment (junghans) [reply core/src/CMakeLists.txt]: @sslattery I did the former, I don't think the latter is necessary.

**Comment by sslattery** [comment]
@dalg24 do you approve?

File: core/src/CMakeLists.txt (hunk 3 above)
Comment (dalg24) [reply core/src/CMakeLists.txt]: > The latter, I was thinking REQUIRED will check for all enabled languages, incl. CXX.

Oh that's right

**Pull request merged**